raiden/storage/serialize.py

```diff
@@ -1,11 +1,78 @@
-import pickle
+import importlib
+import json
+from raiden.utils.serialization import ReferenceCache
 
 
-class PickleSerializer:
+class RaidenJSONEncoder(json.JSONEncoder):
+    """ A custom JSON encoder to provide convenience
+    of recursive instance encoding. """
+    def default(self, obj):
+        """
+        If an object has `to_dict` method,
+        call that method.
+        """
+        if hasattr(obj, 'to_dict'):
+            result = obj.to_dict()
+            result['_type'] = f'{obj.__module__}.{obj.__class__.__name__}'
+            result['_version'] = 0
+            return result
+        return super().default(obj)
+
+
+class RaidenJSONDecoder(json.JSONDecoder):
+    """ A custom JSON decoder which facilitates
+    specific object type invocation to restore
+    it's state.
+    """
+
+    # Maintain a global cache instance
+    # which lives across different decoder instances
+    _ref_cache = ReferenceCache()
```

Comment:
The idea is that we cache deserialized objects at startup so that a given instance of (example) `NettingChannelState` with the exact same attribute values is shared. Because we keep references of the same object across many places so that's why we cache.
The cache should probably not be used after bootstrap is done, will implement this.